File: EOCV-Sim/src/main/java/com/github/serivesmejia/eocvsim/input/source/CameraSource.java
/*
 * Copyright (c) 2021 Sebastian Erives
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 *
 */

package com.github.serivesmejia.eocvsim.input.source;

import com.github.serivesmejia.eocvsim.gui.Visualizer;
import com.github.serivesmejia.eocvsim.input.InputSource;
import com.github.serivesmejia.eocvsim.util.StrUtil;
import com.google.gson.annotations.Expose;
import io.github.deltacv.steve.Webcam;
import io.github.deltacv.steve.WebcamRotation;
import io.github.deltacv.steve.opencv.OpenCvWebcam;
import io.github.deltacv.steve.opencv.OpenCvWebcamBackend;
import io.github.deltacv.steve.openimaj.OpenIMAJWebcamBackend;
import org.opencv.core.Mat;
import org.opencv.core.Size;
import org.opencv.imgproc.Imgproc;
import org.openftc.easyopencv.MatRecycler;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.swing.filechooser.FileFilter;
import java.util.List;

public class CameraSource extends InputSource {

    // for global use, -1 means no webcam currently in use
    public static int currentWebcamIndex = -1;

    protected int webcamIndex;

    @Expose
    protected String webcamName = null;

    private transient Webcam camera = null;

    private transient MatRecycler.RecyclableMat lastFramePaused = null;
    private transient MatRecycler.RecyclableMat lastFrame = null;

    private transient boolean initialized = false;

    protected boolean isLegacyByIndex = false;

    @Expose
    protected volatile Size size;
    @Expose
    protected volatile WebcamRotation rotation;

    private volatile transient MatRecycler matRecycler;

    private transient long capTimeNanos = 0;

    Logger logger = LoggerFactory.getLogger(getClass());

    public CameraSource() {}

    public CameraSource(String webcamName, Size size, WebcamRotation rotation) {
        this.webcamName = webcamName;
        this.size = size;
        this.rotation = rotation;
    }

    public CameraSource(String webcamName, Size size) {
        this.webcamName = webcamName;
        this.size = size;
        this.rotation = WebcamRotation.UPRIGHT;
    }

    public CameraSource(int webcamIndex, Size size, WebcamRotation rotation) {
        this.webcamIndex = webcamIndex;
        this.size = size;
        this.rotation = rotation;
        isLegacyByIndex = true;
    }

    public CameraSource(int webcamIndex, Size size) {
        this.webcamIndex = webcamIndex;
        this.size = size;
        this.rotation = WebcamRotation.UPRIGHT;
        isLegacyByIndex = true;
    }

    @Override
    public boolean init() {
        if (initialized) return false;
        initialized = true;

        if(rotation == null) rotation = WebcamRotation.UPRIGHT;

        if(webcamName != null) {
            Webcam.Companion.setBackend(OpenIMAJWebcamBackend.INSTANCE);

            List<Webcam> webcams = Webcam.Companion.getAvailableWebcams();

            boolean foundWebcam = false;

            for(Webcam device : webcams) {
                String name = device.getName();
                double similarity = StrUtil.similarity(name, webcamName);

                if(name.equals(webcamName) || similarity > 0.6) {
                    logger.info("\"" + name + "\" compared to \"" + webcamName + "\", similarity " + similarity);

                    camera = device;
                    foundWebcam = true;
                    break;
                }
            }

            if(!foundWebcam) {
                logger.error("Could not find webcam " + webcamName);
                return false;
            }
        } else {
            Webcam.Companion.setBackend(OpenCvWebcamBackend.INSTANCE);
            camera = new OpenCvWebcam(webcamIndex, size, rotation);
        }

        camera.setResolution(size);
        camera.setRotation(rotation);
        camera.open();

        if (!camera.isOpen()) {
            logger.error("Unable to open camera " + webcamIndex);
            return false;
        }

        if (matRecycler == null) matRecycler = new MatRecycler(4);
        MatRecycler.RecyclableMat newFrame = matRecycler.takeMat();

        camera.read(newFrame);

        if (newFrame.empty()) {
            logger.error("Unable to open camera " + webcamIndex + ", returned Mat was empty.");
            newFrame.release();
            return false;
        }

        matRecycler.returnMat(newFrame);

        currentWebcamIndex = webcamIndex;

        return true;
    }

    @Override
    public void reset() {
        if (!initialized) return;
        if (camera != null && camera.isOpen()) camera.close();

        if(lastFrame != null && lastFrame.isCheckedOut())
            lastFrame.returnMat();
        if(lastFramePaused != null && lastFramePaused.isCheckedOut())
            lastFramePaused.returnMat();

        initialized = false;
    }

    @Override
    public void close() {
        if (camera != null && camera.isOpen()) camera.close();
        currentWebcamIndex = -1;
    }

    @Override
    public Mat update() {
        if (isPaused) {
            return lastFramePaused;
        } else if (lastFramePaused != null) {
            lastFramePaused.release();
            lastFramePaused.returnMat();
            lastFramePaused = null;
        }

        if (lastFrame == null) lastFrame = matRecycler.takeMat();
        if (camera == null) return lastFrame;

        MatRecycler.RecyclableMat newFrame = matRecycler.takeMat();

        camera.read(newFrame);
        capTimeNanos = System.nanoTime();

        if (newFrame.empty()) {
            newFrame.returnMat();
            return lastFrame;
        }

        if (size == null) size = lastFrame.size();

        // not needed. cameras return RGB now
        // Imgproc.cvtColor(newFrame, lastFrame, Imgproc.COLOR_BGR2RGB);
        newFrame.copyTo(lastFrame);

        newFrame.release();
        newFrame.returnMat();

        return lastFrame;
    }

    @Override
    public void onPause() {
        if (lastFrame != null) lastFrame.release();
        if (lastFramePaused == null) lastFramePaused = matRecycler.takeMat();

        camera.read(lastFramePaused);
        Imgproc.cvtColor(lastFramePaused, lastFramePaused, Imgproc.COLOR_BGR2RGB);

        update();

        camera.close();

        currentWebcamIndex = -1;
    }

    @Override
    public void onResume() {
        Visualizer.AsyncPleaseWaitDialog apwdCam = eocvSim.inputSourceManager.showApwdIfNeeded(name);
        camera.open();

        apwdCam.destroyDialog();
    }

    @Override
    protected InputSource internalCloneSource() {
        return new CameraSource(webcamName, size);
    }

    @Override
    public FileFilter getFileFilters() {
        return null;
    }

    @Override
    public long getCaptureTimeNanos() {
        return capTimeNanos;
    }

    @Override
    public String toString() {
        if (size == null) size = new Size();
        return "CameraSource(" + webcamName + ", " + webcamIndex + ", " + (size != null ? size.toString() : "null") + ")";
    }

}

File: EOCV-Sim/src/main/java/com/github/serivesmejia/eocvsim/input/source/CameraSource.kt
/*
 * Copyright (c) 2026 Sebastian Erives
 * Licensed under the MIT License.
 */

package com.github.serivesmejia.eocvsim.input.source

import com.fasterxml.jackson.annotation.JsonAutoDetect
import com.fasterxml.jackson.annotation.JsonProperty
import com.github.serivesmejia.eocvsim.input.InputSource
import com.github.serivesmejia.eocvsim.input.InputSourceInitializer
import com.github.serivesmejia.eocvsim.config.ConfigManager
import io.github.deltacv.common.util.loggerForThis
import org.koin.core.component.KoinComponent
import org.koin.core.component.inject
import org.opencv.core.Mat
import org.opencv.core.Size
import org.opencv.imgproc.Imgproc
import org.wpilib.util.PixelFormat
import org.wpilib.vision.camera.CvSink
import org.wpilib.vision.camera.UsbCamera
import org.wpilib.vision.camera.VideoMode
import javax.swing.filechooser.FileFilter

@JsonAutoDetect(
    fieldVisibility = JsonAutoDetect.Visibility.NONE,
    getterVisibility = JsonAutoDetect.Visibility.NONE,
    isGetterVisibility = JsonAutoDetect.Visibility.NONE,
    setterVisibility = JsonAutoDetect.Visibility.NONE,
    creatorVisibility = JsonAutoDetect.Visibility.NONE
)
class CameraSource : InputSource, KoinComponent {

    companion object {
        @JvmStatic var currentWebcamIndex = -1
    }

    override val hasSlowInitialization = true

    @JsonProperty @JvmField var cameraPortIndex = -1
    @JsonProperty @JvmField var exactPortMatch = false
    @JsonProperty @JvmField var vendorId: Int? = null
    @JsonProperty @JvmField var productId: Int? = null
    @Transient var webcamName = ""

    @JsonProperty @JvmField var videoMode: VideoMode? = null

    @Transient var camera: UsbCamera? = null
        private set

    @Transient private var cvSink: CvSink? = null
    @Transient private var lastFrame = Mat()
    @Transient private var initialized = false
    @Transient var isLegacyByIndex = false
    @Transient private var capTimeNanos = 0L

    private val configManager: ConfigManager by inject()
    private val logger by loggerForThis()

    constructor() : super()

    constructor(webcamName: String?, videoMode: VideoMode?) : super() {
        this.webcamName = webcamName ?: ""
        this.videoMode = videoMode
    }

    constructor(webcamName: String?, vendorId: Int?, productId: Int?, videoMode: VideoMode?) : super() {
        this.webcamName = webcamName ?: ""
        this.vendorId = vendorId
        this.productId = productId
        this.videoMode = videoMode
    }

    constructor(
        webcamName: String?,
        cameraPortIndex: Int,
        exactPortMatch: Boolean,
        vendorId: Int?,
        productId: Int?,
        videoMode: VideoMode?
    ) : super() {
        this.webcamName = webcamName ?: ""
        this.cameraPortIndex = cameraPortIndex
        this.exactPortMatch = exactPortMatch
        this.vendorId = vendorId
        this.productId = productId
        this.videoMode = videoMode
    }

    constructor(cameraPortIndex: Int, videoMode: VideoMode?) : super() {
        this.cameraPortIndex = cameraPortIndex
        this.videoMode = videoMode
        this.isLegacyByIndex = true
    }

    override val sourceSize: Size
        get() = videoMode?.let { Size(it.width.toDouble(), it.height.toDouble()) } ?: Size()

    override fun init(): Boolean {
        if (initialized) return false
        initialized = true

        val cameras by lazy { UsbCamera.enumerateUsbCameras() }

        val matchedInfo = when {
            exactPortMatch -> cameras.firstOrNull {
                cameraPortIndex >= 0 &&
                        it.dev == cameraPortIndex &&
                        vendorId != null && productId != null &&
                        it.vendorId == vendorId &&
                        it.productId == productId
            } ?: run {
                logger.error("Camera not found on the same connection: $cameraPortIndex")
                return false
            }

            cameraPortIndex >= 0 -> cameras.firstOrNull { it.dev == cameraPortIndex }
                ?: run {
                    logger.error("Camera not found on port: $cameraPortIndex")
                    return false
                }

            vendorId != null && productId != null -> cameras.firstOrNull {
                it.vendorId == vendorId && it.productId == productId
            } ?: run {
                logger.error("Camera not found by VID/PID: $vendorId:$productId")
                return false
            }

            webcamName.isNotEmpty() -> cameras.firstOrNull { it.name == webcamName }
                ?: run {
                    logger.error("Camera not found: $webcamName")
                    return false
                }

            else -> null
        }

        camera = if (matchedInfo != null) {
            webcamName = matchedInfo.name
            UsbCamera(matchedInfo.name, matchedInfo.dev)
        } else {
            UsbCamera("$cameraPortIndex", cameraPortIndex)
        }

        camera!!.videoMode = videoMode ?: camera!!.videoMode

        logger.info("Camera started: ${matchedInfo?.name ?: webcamName.ifEmpty { "Camera $cameraPortIndex" }} ${camera!!.videoMode?.stringify()}")

        cvSink = CvSink("eocvsim_sink_$cameraPortIndex", PixelFormat.BGR).also {
            it.source = camera
        }

        if (cvSink!!.grabFrame(lastFrame, configManager.config.webcamOpenTimeoutSec) == 0L || lastFrame.empty()) {
            logger.error("Failed to open camera: ${cvSink!!.error}")
            return false
        }

        currentWebcamIndex = cameraPortIndex
        return true
    }

    override fun reset() {
        if (!initialized) return
        teardown()
        lastFrame.release()
        initialized = false
    }

    override fun close() = teardown()

    override fun update(): Mat {
        if (isPaused) return lastFrame

        capTimeNanos = cvSink?.grabFrame(lastFrame, configManager.config.webcamNewFrameTimeoutSec) ?: 0L

        if (!lastFrame.empty()) {
            Imgproc.cvtColor(lastFrame, lastFrame, Imgproc.COLOR_BGR2RGBA)
        }

        return lastFrame
    }

    override fun onPause() {
        cvSink?.grabFrame(lastFrame, configManager.config.webcamNewFrameTimeoutSec)
        teardown()
    }

    override fun onResume() {
        InputSourceInitializer.runWithTimeout(this) { init() }
    }

    private fun teardown() {
        cvSink?.close()
        cvSink = null
        camera?.close()
        camera = null
        currentWebcamIndex = -1
    }

    override fun internalCloneSource(): InputSource =
        if (isLegacyByIndex) CameraSource(cameraPortIndex, videoMode)
        else CameraSource(webcamName, cameraPortIndex, exactPortMatch, vendorId, productId, videoMode)

    override val fileFilters: FileFilter? = null
    override val captureTimeNanos get() = capTimeNanos

    override fun toString() =
        "CameraSource($webcamName, port=$cameraPortIndex, exactPortMatch=$exactPortMatch, vid=$vendorId, pid=$productId, ${videoMode?.stringify()})"

    private fun VideoMode.stringify() = "${width}x${height}@${fps} $pixelFormat"
}